ROS2: Image proc non zero crop

klintan opened the pull request:
As per issue https://github.com/ros-perception/image_pipeline/issues/430 there are some missing nodes from image_proc package that needs to be added.

This is for the CropNonZeroNode.

It is built very very similar to the Rectify and Debayer node, so maybe review should be conjunction with https://github.com/ros-perception/image_pipeline/pull/457 .

klintan committed:
added crop non zero composable node

SteveMacenski commented on image_proc/src/crop_non_zero.cpp:
```diff
@@ -0,0 +1,154 @@
+
+    // search the largest area
+    /* // -std=c++11
+  std::vector<std::vector<cv::Point> >::iterator it = std::max_element(cnt.begin(), cnt.end(), [](std::vector<cv::Point> a, std::vector<cv::Point> b) {
```
commented out?

klintan replied:
Used this and remvoved the other since this is the valid one for c++11 or newer.

SteveMacenski commented on image_proc/src/crop_non_zero.cpp:
```diff
@@ -0,0 +1,154 @@
+    std::vector<std::vector<cv::Point>>::iterator it = cnt.begin();
+    for (std::vector<std::vector<cv::Point>>::iterator i = cnt.begin(); i != cnt.end(); ++i)
+    {
+        it = (*it).size() < (*i).size() ? i : it;
```
Can you change this all to 2 spaces per indent?

klintan replied:
Will fix!

klintan replied:
Fixed everything to be 2 spaces indentation.

SteveMacenski reviewed (changes requested):
LGTM, is this just a straight port? Anything you want to mention?

klintan commented:
> LGTM, is this just a straight port? Anything you want to mention?

Nothing in particular, I need to do some more testing. But should be a straight port.

SteveMacenski commented:
OK tell me when you're happy with it after testing all the relevent options

klintan committed:
fixed indentation and uncommented std11 code

klintan committed:
fixed some missed issues and tested

klintan commented:
Rebased on top of the new master and fixed PR feedback. Fixed some other issues I forgot to handle when testing it. I couldn't find any issues now at least, after testing. 

Let me know if I missed something or if you have any feedback on my changes.

SteveMacenski commented on image_proc/CMakeLists.txt:
```diff
@@ -76,6 +76,25 @@ ament_target_dependencies(resize
+target_compile_definitions(crop_non_zero
+  PRIVATE "COMPOSITION_BUILDING_DLL")
+
+ament_target_dependencies(crop_non_zero
```
Har har har, `${dependencies}` variable comes back :-)

SteveMacenski reviewed (changes requested):
Small stuff but looks fine

klintan committed:
fixed small miss dependencies variable

klintan committed:
messed up, typo

klintan commented:
Tests failing on  `image_publisher` not sure why, perhaps a temporary Circle issue, I don't have rights to rerun the workflow either. Should not affect this code.

SteveMacenski commented on image_proc/src/crop_non_zero.cpp:
```diff
@@ -0,0 +1,149 @@
+#include <boost/thread.hpp>
+#include <cv_bridge/cv_bridge.h>
+#include "crop_non_zero.hpp"
+//#include <algorithm> // for std::max_element
```
Commented out?

SteveMacenski replied:
You use this below

klintan replied:
thanks, forgot to uncomment when I switched to this other version requiring this one. Not sure why the compiler did not complain.

SteveMacenski commented on image_proc/include/crop_non_zero.hpp:
```diff
@@ -0,0 +1,69 @@
+    CropNonZeroNode(const rclcpp::NodeOptions&);
+  private:
+    std::string camera_namespace_;
+    std::string image_pub_;
```
Can you rename image_pub / sub? I assumed those were actually pub/subs and not the string names. That offered a bit of confusion

SteveMacenski commented on image_proc/include/crop_non_zero.hpp:
```diff
@@ -0,0 +1,69 @@
+    std::string image_sub_;
+
+    // Subscriptions
+    image_transport::Subscriber sub_raw_;
```
"raw"?

klintan replied:
Because of the compression type (also the name ROS1 uses). But I'm open to changing.

SteveMacenski replied:
Ah ok. Thats fine

SteveMacenski commented on image_proc/src/crop_non_zero.cpp:
```diff
@@ -0,0 +1,149 @@
+  std::lock_guard<std::mutex> lock(connect_mutex_);
+
+  RCLCPP_INFO(this->get_logger(), "subscribe: %s", image_sub_.c_str());
+  sub_raw_ = image_transport::create_subscription(this, image_sub_, std::bind(&CropNonZeroNode::imageCb, this, std::placeholders::_1), "raw");
```
Should the type of compression be a parameter? (actually asking, I'm not sure what the ROS1 version did)

klintan replied:
Seems it's just hard-coded: 

```    
image_transport::TransportHints hints("raw", ros::TransportHints(), getPrivateNodeHandle());

sub_raw_ = it_->subscribe("image_raw", 1, &CropNonZeroNodelet::imageCb, this, hints);
```

SteveMacenski replied:
Eh. Not ideal but not going to make it part of this PR

SteveMacenski commented on image_proc/src/crop_non_zero.cpp:
```diff
@@ -0,0 +1,149 @@
+}
+
+// Handles (un)subscribing when clients (un)subscribe
+void CropNonZeroNode::connectCb()
```
So connectCb is only called in this lambda, and the next line in the lambda relocks the same mutex. It would be better to either:

a) put the connectCb stuff  inside of the lambda, if there's nothing being otherwise exposed from the outside that needs it
b) only have the mutex in the lambda before the connectCb is called so we're not having 2 locks toggling quickly

klintan replied:
Went for b) (thinking there was some logic for remove a subscription when the subscriber number were 0 that might be added in the future)

klintan committed:
PR feedback, topic names and mutex lock

SteveMacenski reviewed (approved):
LGTM 

just needs circle ci building

SteveMacenski commented:
@klintan pull in ros2 once I have #470 merged. This will fix builds so it should just be whatever changes you've made

SteveMacenski commented on image_proc/include/crop_non_zero.hpp:
```diff
@@ -0,0 +1,69 @@
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+#ifndef IMAGE_PROC_PROCESSOR_H
```
You need to update your guards

klintan replied:
Done

JWhitleyWork commented:
CI restarted. The error that came up is usually because the compiler ran out of memory.

klintan committed:
updated header guard

klintan committed:
from jwhitley review, one argument constructors should be explicit

klintan commented:
> @klintan pull in ros2 once I have #470 merged. This will fix builds so it should just be whatever changes you've made

Hm I did a rebase on the latest changes, but still failing :/

JWhitleyWork commented:
@klintan - I'm testing this locally to see what's up. It's a compiler error but I'm not sure why it's happening yet.

JWhitleyWork commented:
@klintan - I created https://github.com/klintan/image_pipeline/pull/3, which I think will solve the problem. Please test it.

SteveMacenski commented:
@JWhitleyAStuff can you PR that and just merge it? That's been irritating me as well

SteveMacenski reviewed (approved):
I'm happy again when @JWhitleyAStuff is

JWhitleyWork commented:
@klintan - Once you rebase on `master`, I'm good to merge.

SteveMacenski commented:
You mean on `ros2`, right?

JWhitleyWork commented:
Yes, sorry.

klintan commented:
i'll rebase asap.

klintan committed:
fixed test issues and small crop_decimate change for compiling

klintan commented:
@JWhitleyAStuff I had to change 
`dst.create(src.rows / 2, src.cols / 2, cv::DataType<DstPixel>::type);`
to
`dst.create(src.rows / 2, src.cols / 2, cv::traits::Type<DstPixel>::value);` 
per https://github.com/opencv/opencv/issues/10115 , which version of opencv do you run? (I'm on 3.4.5) otherwise it wouldn't build. Perhaps it's something with my env since circleci seems to be fine with that ? 

Fixed all other issues.  Perhaps the crop_decimate fix should be in a different PR if it's a problem.

edit: circleci does not build with that change, so reverted regardless. But I think it's worth looking into that issue.

klintan committed:
include file changed and reverted crop decimate change

SteveMacenski commented:
@klintan that's a good use for the preprocessor `if`s on the version, there's a bunch of examples in the code for supporting multiple opencv versions that way

klintan committed:
preprocessor condition for different opencv version

klintan committed:
fixed linting issues

klintan commented:
> @klintan that's a good use for the preprocessor `if`s on the version, there's a bunch of examples in the code for supporting multiple opencv versions that way


Ok used a condition for the opencv version, that fixed the issue on circle and locally.

SteveMacenski commented on image_proc/src/crop_decimate.cpp:
```diff
@@ -46,7 +46,11 @@ void debayer2x2toBGR(
 {
   typedef cv::Vec<T, 3> DstPixel;  // 8- or 16-bit BGR
-  dst.create(src.rows / 2, src.cols / 2, cv::DataType<DstPixel>::type);
+  #if CV_MAJOR_VERSION >= 3 && CV_MINOR_VERSION > 2
```
Does the minor version actually matter? just curious.

klintan replied:
Yes, or put another way; the change was made in 3.3.1 it seems, https://github.com/ros-perception/image_pipeline/pull/302

edit: to be safe, I guess we would need a CV_SUBMINOR_VERSION as well ? 3.3.0 would use the new way as well. What do you think?

SteveMacenski replied:
I mean I think this is fine

SteveMacenski commented:
Just waiting on @JWhitleyAStuff feedback then

JWhitleyWork reviewed (approved):
LGTM!

SteveMacenski commented:
@klintan any thing else you want to add before merging?
